File: prac3-6/lista.py
class Nodo:
    def __init__(self, dato):
        self.dato = dato
        self.siguiente = None


class ListaEnlazada:
    def __init__(self):
        self.cabeza = None

    def insertar_inicio(self, dato):
        nuevo = Nodo(dato)
        nuevo.siguiente = self.cabeza
        self.cabeza = nuevo

    def insertar_final(self, dato):
        nuevo = Nodo(dato)

        if not self.cabeza:
            self.cabeza = nuevo
            return

        actual = self.cabeza
        while actual.siguiente:
            actual = actual.siguiente

        actual.siguiente = nuevo

    def eliminar_inicio(self):
        if self.cabeza:
            self.cabeza = self.cabeza.siguiente

    def mostrar(self):
        actual = self.cabeza
        while actual:
            print(actual.dato, end=" -> ")
            actual = actual.siguiente
        print("None")


lista = ListaEnlazada()

lista.insertar_inicio(8)
lista.insertar_inicio(4)
lista.insertar_final(11)

print("Lista inicial:")
lista.mostrar()

lista.eliminar_inicio()

print("Lista final:")
lista.mostrar()

File: unidad-1/prac3-6/lista.cpp
#include <iostream>
using namespace std;

struct Nodo {
    int dato;
    Nodo* siguiente;
};

class Lista {
private:
    Nodo* cabeza;

public:
    Lista() {
        cabeza = nullptr;
    }

    void insertarInicio(int valor) {
        Nodo* nuevo = new Nodo();
        nuevo->dato = valor;
        nuevo->siguiente = cabeza;
        cabeza = nuevo;
    }

    void insertarFinal(int valor) {
        Nodo* nuevo = new Nodo();
        nuevo->dato = valor;
        nuevo->siguiente = nullptr;

        if (cabeza == nullptr) {
            cabeza = nuevo;
            return;
        }

        Nodo* temp = cabeza;

        while (temp->siguiente != nullptr) {
            temp = temp->siguiente;
        }

        temp->siguiente = nuevo;
    }

    void eliminarInicio() {
        if (cabeza != nullptr) {
            Nodo* temp = cabeza;
            cabeza = cabeza->siguiente;
            delete temp;
        }
    }

    void mostrar() {
        Nodo* temp = cabeza;

        while (temp != nullptr) {
            cout << temp->dato << " -> ";
            temp = temp->siguiente;
        }

        cout << "NULL" << endl;
    }
};

int main() {
    Lista lista;

    lista.insertarInicio(8);
    lista.insertarInicio(4);
    lista.insertarFinal(11);

    lista.eliminarInicio();

    lista.mostrar();

    return 0;
}